File: cortex/autopilot/policies.py
"""cortex.autopilot.policies — Consolidated policy layer for the Autopilot module.

Replaces the previous ``cortex.autopilot.policies.{base,default,auto_checkpoint}``
subpackage with a single module designed for the post-Phase-03 architecture:
**Autopilot is a thin policy + hook layer over the cortex.session primitive.**

Public API:
    :class:`AutopilotMode`        — observe / assist / autopilot
    :class:`AutopilotPolicy`      — frozen dataclass: mode + thresholds + flags
    :class:`EnforcementSeverity`  — proceed / warn / block
    :class:`EnforcementResult`    — frozen dataclass: severity + reason
    :class:`PolicyEnforcer`       — applies the policy at lifecycle hooks

The enforcer's hooks return a list of :class:`EnforcementResult` so the caller
(typically :class:`cortex.autopilot.service.AutopilotService`) decides how to
surface warnings or stop the operation on a block.

Design note:
    All time comparisons use timezone-aware UTC. The enforcer is stateless
    beyond its immutable policy, and pure functions of the inputs.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from enum import StrEnum

from cortex.autopilot.config import AutopilotConfig
from cortex.session.models import Checkpoint, SessionRecord


class AutopilotMode(StrEnum):
    """Operational stance of the Autopilot layer for a session.

    - ``OBSERVE``   — hooks register checkpoints; no warnings, no interventions.
    - ``ASSIST``    — observe + warnings (out-of-scope, stalled work, security).
    - ``AUTOPILOT`` — assist + active interventions (block on missing
      verification, etc.).
    """

    OBSERVE = "observe"
    ASSIST = "assist"
    AUTOPILOT = "autopilot"


class EnforcementSeverity(StrEnum):
    """How strongly a policy enforcement signals its decision."""

    PROCEED = "proceed"
    WARN = "warn"
    BLOCK = "block"


@dataclass(frozen=True)
class EnforcementResult:
    """Outcome of one policy rule applied at a hook.

    A hook may return multiple results (one per rule). The caller is
    responsible for combining them — usually by aggregating WARNs and
    raising / aborting on the first BLOCK.
    """

    severity: EnforcementSeverity
    reason: str = ""

    @property
    def allowed(self) -> bool:
        """True unless severity is BLOCK."""
        return self.severity is not EnforcementSeverity.BLOCK

    @classmethod
    def proceed(cls) -> EnforcementResult:
        return cls(severity=EnforcementSeverity.PROCEED, reason="")

    @classmethod
    def warn(cls, reason: str) -> EnforcementResult:
        return cls(severity=EnforcementSeverity.WARN, reason=reason)

    @classmethod
    def block(cls, reason: str) -> EnforcementResult:
        return cls(severity=EnforcementSeverity.BLOCK, reason=reason)


# Budget-profile registry — single source of truth (formerly
# cortex.autopilot.context_budget.BUDGET_PROFILES, kept here so the policy
# layer has no cross-module data dependency).
DEFAULT_BUDGET_PROFILE = "fast_code"
KNOWN_BUDGET_PROFILES: frozenset[str] = frozenset(
    {"question_only", "docs_only", "fast_code", "deep_code", "finish_only"}
)


# Security keyword set reused from the old SecuritySensitiveDetector. Drives
# the on_session_open warning when a spec summary mentions risky topics.
_SECURITY_KEYWORDS = frozenset(
    {
        "password",
        "secret",
        "token",
        "jwt",
        "oauth",
        "auth",
        "login",
        "permission",
        "role",
        "rbac",
        "acl",
        "crypto",
        "encrypt",
        "decrypt",
        "hash",
        "salt",
    }
)
_SECURITY_KEYWORD_PATTERN = re.compile(
    r"\b(" + "|".join(re.escape(k) for k in _SECURITY_KEYWORDS) + r")\b",
    re.IGNORECASE,
)


@dataclass(frozen=True)
class AutopilotPolicy:
    """Declarative policy applied to a Session under Autopilot supervision.

    The policy is immutable per session. The caller selects a mode at
    start-time and :class:`PolicyEnforcer` evaluates the same instance at
    every lifecycle hook.

    Attributes:
        mode: Operational stance. See :class:`AutopilotMode`.
        budget_profile: Name of the context-budget profile from
            :data:`KNOWN_BUDGET_PROFILES`. Documentational; the actual
            budget enforcement lives in ``cortex.context_enricher``.
        pre_commit_verification: When True (effective only in AUTOPILOT
            mode), :meth:`PolicyEnforcer.on_pre_close` blocks unless at
            least one checkpoint carries verified claims.
        out_of_scope_warning: When True (effective in ASSIST and
            AUTOPILOT), :meth:`PolicyEnforcer.on_checkpoint` warns if the
            new checkpoint touches files outside the spec's scope.
        auto_checkpoint_threshold_files: After N artifact paths touched
            without an intervening verified checkpoint, the enforcer
            nudges the user with a warning.
        auto_checkpoint_threshold_minutes: After M minutes since the prior
            checkpoint (when the new one touches any files), the enforcer
            nudges the user with a warning.
        warn_on_security_summary: When True (effective in ASSIST and
            AUTOPILOT), :meth:`PolicyEnforcer.on_session_open` warns if
            the spec summary mentions security-sensitive terms.
    """

    mode: AutopilotMode = AutopilotMode.ASSIST
    budget_profile: str = DEFAULT_BUDGET_PROFILE
    pre_commit_verification: bool = False
    out_of_scope_warning: bool = True
    auto_checkpoint_threshold_files: int = 5
    auto_checkpoint_threshold_minutes: int = 10
    warn_on_security_summary: bool = True

    def __post_init__(self) -> None:
        if self.budget_profile not in KNOWN_BUDGET_PROFILES:
            raise ValueError(
                f"unknown budget_profile {self.budget_profile!r}; "
                f"must be one of {sorted(KNOWN_BUDGET_PROFILES)}"
            )
        if self.auto_checkpoint_threshold_files < 1:
            raise ValueError(
                "auto_checkpoint_threshold_files must be >= 1, "
                f"got {self.auto_checkpoint_threshold_files}"
            )
        if self.auto_checkpoint_threshold_minutes < 1:
            raise ValueError(
                "auto_checkpoint_threshold_minutes must be >= 1, "
                f"got {self.auto_checkpoint_threshold_minutes}"
            )

    @classmethod
    def from_config(cls, config: AutopilotConfig) -> AutopilotPolicy:
        """Build a policy from an :class:`AutopilotConfig` (YAML-backed).

        Unknown ``config.mode`` or ``config.default_budget_profile`` values
        fall back to safe defaults so a typo in ``autopilot.yaml`` never
        crashes the CLI. The mistake is surfaced by
        :mod:`cortex.autopilot.doctor`.

        Higher modes default to more enforcement:
            OBSERVE    — all warnings off, no pre-commit verification.
            ASSIST     — warnings on, no pre-commit verification.
            AUTOPILOT  — warnings on, pre-commit verification on.
        """
        try:
            mode = AutopilotMode(config.mode)
        except ValueError:
            mode = AutopilotMode.ASSIST

        budget = (
            config.default_budget_profile
            if config.default_budget_profile in KNOWN_BUDGET_PROFILES
            else DEFAULT_BUDGET_PROFILE
        )

        is_observe = mode is AutopilotMode.OBSERVE
        return cls(
            mode=mode,
            budget_profile=budget,
            pre_commit_verification=mode is AutopilotMode.AUTOPILOT,
            out_of_scope_warning=not is_observe,
            warn_on_security_summary=not is_observe,
            auto_checkpoint_threshold_files=max(1, config.auto_checkpoint_files),
            auto_checkpoint_threshold_minutes=max(1, config.auto_checkpoint_minutes),
        )


class PolicyEnforcer:
    """Apply an :class:`AutopilotPolicy` at lifecycle hooks of a SessionRecord.

    Stateless beyond the immutable policy. All time comparisons use UTC.
    """

    def __init__(self, policy: AutopilotPolicy) -> None:
        self._policy = policy

    @property
    def policy(self) -> AutopilotPolicy:
        return self._policy

    # ── Hook: on_session_open ─────────────────────────────────────

    def on_session_open(
        self,
        session: SessionRecord,
        *,
        spec_summary: str | None = None,
    ) -> list[EnforcementResult]:
        """Pre-check after the session is opened.

        Warnings issued:
            - Spec summary mentions security-sensitive terms (when the
              policy has ``warn_on_security_summary``).
        """
        results: list[EnforcementResult] = []
        if self._policy.mode is AutopilotMode.OBSERVE:
            return results

        summary = spec_summary if spec_summary is not None else session.spec_summary
        if self._policy.warn_on_security_summary and _looks_security_sensitive(summary):
            results.append(
                EnforcementResult.warn(
                    "Spec summary mentions security-sensitive terms — review the diff "
                    "carefully before closing the session."
                )
            )
        return results

    # ── Hook: on_checkpoint ───────────────────────────────────────

    def on_checkpoint(
        self,
        session: SessionRecord,
        checkpoint: Checkpoint,
        *,
        files_in_scope: list[str] | None = None,
    ) -> list[EnforcementResult]:
        """Evaluate after a checkpoint is appended.

        Warnings issued:
            - WARN if ``out_of_scope_warning`` is set, the spec declares
              ``files_in_scope``, and the checkpoint touches paths outside
              that scope.
            - WARN if more than ``auto_checkpoint_threshold_files`` distinct
              artifact paths have been touched since the last checkpoint
              with verified claims.
            - WARN if more than ``auto_checkpoint_threshold_minutes`` have
              passed since the previous checkpoint AND the new one touches
              any files.
        """
        results: list[EnforcementResult] = []
        if self._policy.mode is AutopilotMode.OBSERVE:
            return results

        if self._policy.out_of_scope_warning and files_in_scope is not None:
            in_scope_set = set(files_in_scope)
            drift = [a for a in checkpoint.artifacts_touched if a not in in_scope_set]
            if drift:
                results.append(
                    EnforcementResult.warn(
                        f"Checkpoint touches files outside spec scope: {sorted(drift)}"
                    )
                )

        files_since_verified = _files_since_last_verified(session)
        if files_since_verified > self._policy.auto_checkpoint_threshold_files:
            results.append(
                EnforcementResult.warn(
                    f"{files_since_verified} artifact paths touched without a "
                    "checkpoint that records verified claims"
                )
            )

        # Time-since-previous-checkpoint test excludes the checkpoint we
        # just appended (``session.checkpoints[-1] is checkpoint``).
        prior_checkpoints = session.checkpoints[:-1] if session.checkpoints else []
        if prior_checkpoints and checkpoint.artifacts_touched:
            elapsed = datetime.now(UTC) - prior_checkpoints[-1].timestamp
            threshold = timedelta(minutes=self._policy.auto_checkpoint_threshold_minutes)
            if elapsed > threshold:
                minutes = int(elapsed.total_seconds() // 60)
                results.append(
                    EnforcementResult.warn(
                        f"{minutes} minutes since the previous checkpoint and the new "
                        "one already touches files — consider checkpointing more often"
                    )
                )
        return results

    # ── Hook: on_pre_close ────────────────────────────────────────

    def on_pre_close(self, session: SessionRecord) -> list[EnforcementResult]:
        """Evaluate before transitioning the session to a terminal status.

        Blocks issued:
            - In AUTOPILOT with ``pre_commit_verification``, BLOCK if no
              checkpoint contains any ``verified_claims``.
        """
        results: list[EnforcementResult] = []
        if (
            self._policy.mode is AutopilotMode.AUTOPILOT
            and self._policy.pre_commit_verification
            and not _has_verified_checkpoint(session)
        ):
            results.append(
                EnforcementResult.block(
                    "Autopilot mode requires at least one checkpoint with verified "
                    "claims before closing the session (set "
                    "pre_commit_verification=False to opt out)."
                )
            )
        return results


# ── Module-private helpers ────────────────────────────────────────


def _looks_security_sensitive(text: str) -> bool:
    if not text:
        return False
    return bool(_SECURITY_KEYWORD_PATTERN.search(text))


def _has_verified_checkpoint(session: SessionRecord) -> bool:
    return any(cp.verified_claims for cp in session.checkpoints)


def _files_since_last_verified(session: SessionRecord) -> int:
    """Distinct artifact paths touched since the last verified checkpoint."""
    touched: set[str] = set()
    for cp in reversed(session.checkpoints):
        if cp.verified_claims:
            break
        touched.update(cp.artifacts_touched)
    return len(touched)


__all__ = [
    "AutopilotMode",
    "AutopilotPolicy",
    "DEFAULT_BUDGET_PROFILE",
    "EnforcementResult",
    "EnforcementSeverity",
    "KNOWN_BUDGET_PROFILES",
    "PolicyEnforcer",
]

File: rust/crates/cortex-autopilot/src/policies.rs
//! Puerto de `cortex.autopilot.policies`: modo, policy inmutable y enforcer
//! de lifecycle hooks. Todos los mensajes contractuales se replican.

use std::sync::Arc;

use chrono::{DateTime, Utc};

use cortex_enterprise::clock::Clock;
use cortex_enterprise::error::EnterpriseError;

use crate::config::AutopilotConfig;
use crate::session_models::{Checkpoint, SessionRecord, SessionStatus};

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AutopilotMode {
    Observe,
    Assist,
    Autopilot,
}

impl AutopilotMode {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Observe => "observe",
            Self::Assist => "assist",
            Self::Autopilot => "autopilot",
        }
    }
    fn parse(s: &str) -> Option<Self> {
        match s {
            "observe" => Some(Self::Observe),
            "assist" => Some(Self::Assist),
            "autopilot" => Some(Self::Autopilot),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EnforcementSeverity {
    Proceed,
    Warn,
    Block,
}

impl PartialEq<&str> for EnforcementSeverity {
    fn eq(&self, other: &&str) -> bool {
        matches!(
            (self, *other),
            (Self::Proceed, "proceed") | (Self::Warn, "warn") | (Self::Block, "block")
        )
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct EnforcementResult {
    pub severity: EnforcementSeverity,
    pub reason: String,
}

impl EnforcementResult {
    pub fn allowed(&self) -> bool {
        self.severity != EnforcementSeverity::Block
    }
    pub fn proceed() -> Self {
        Self {
            severity: EnforcementSeverity::Proceed,
            reason: String::new(),
        }
    }
    pub fn warn(reason: impl Into<String>) -> Self {
        Self {
            severity: EnforcementSeverity::Warn,
            reason: reason.into(),
        }
    }
    pub fn block(reason: impl Into<String>) -> Self {
        Self {
            severity: EnforcementSeverity::Block,
            reason: reason.into(),
        }
    }
}

pub const DEFAULT_BUDGET_PROFILE: &str = "fast_code";
pub const KNOWN_BUDGET_PROFILES: &[&str] = &[
    "question_only",
    "docs_only",
    "fast_code",
    "deep_code",
    "finish_only",
];

const SECURITY_KEYWORD_PATTERN_WORDS: &[&str] = &[
    "password",
    "secret",
    "token",
    "jwt",
    "oauth",
    "auth",
    "login",
    "permission",
    "role",
    "rbac",
    "acl",
    "crypto",
    "encrypt",
    "decrypt",
    "hash",
    "salt",
];

fn looks_security_sensitive(text: &str) -> bool {
    if text.is_empty() {
        return false;
    }
    // \b(palabra)\b case-insensitive — replica el regex de Python.
    let lower = format!(" {text} ").to_lowercase();
    for kw in SECURITY_KEYWORD_PATTERN_WORDS {
        let mut from = 0usize;
        while let Some(pos) = lower[from..].find(kw) {
            let start = from + pos;
            let end = start + kw.len();
            let before = lower[..start].chars().next_back();
            let after = lower[end..].chars().next();
            let boundary_before = before
                .map(|c| !(c.is_alphanumeric() || c == '_'))
                .unwrap_or(true);
            let boundary_after = after
                .map(|c| !(c.is_alphanumeric() || c == '_'))
                .unwrap_or(true);
            if boundary_before && boundary_after {
                return true;
            }
            from = start + 1;
        }
    }
    false
}

fn has_verified_checkpoint(session: &SessionRecord) -> bool {
    session
        .checkpoints
        .iter()
        .any(|cp| !cp.verified_claims.is_empty())
}

fn files_since_last_verified(session: &SessionRecord) -> usize {
    let mut touched = std::collections::BTreeSet::new();
    for cp in session.checkpoints.iter().rev() {
        if !cp.verified_claims.is_empty() {
            break;
        }
        touched.extend(cp.artifacts_touched.iter().cloned());
    }
    touched.len()
}

#[derive(Debug, Clone)]
pub struct AutopilotPolicy {
    pub mode: AutopilotMode,
    pub budget_profile: String,
    pub pre_commit_verification: bool,
    pub out_of_scope_warning: bool,
    pub auto_checkpoint_threshold_files: i64,
    pub auto_checkpoint_threshold_minutes: i64,
    pub warn_on_security_summary: bool,
}

impl AutopilotPolicy {
    /// Construcción directa con validaciones de `__post_init__`.
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        mode: AutopilotMode,
        budget_profile: String,
        pre_commit_verification: bool,
        out_of_scope_warning: bool,
        auto_checkpoint_threshold_files: i64,
        auto_checkpoint_threshold_minutes: i64,
        _clock: Arc<dyn Clock>,
    ) -> Result<Self, EnterpriseError> {
        if !KNOWN_BUDGET_PROFILES.contains(&budget_profile.as_str()) {
            let mut sorted: Vec<&str> = KNOWN_BUDGET_PROFILES.to_vec();
            sorted.sort_unstable();
            return Err(EnterpriseError::Validation(format!(
                "unknown budget_profile '{}'; must be one of [{}]",
                budget_profile,
                sorted
                    .iter()
                    .map(|s| format!("{s:?}"))
                    .collect::<Vec<_>>()
                    .join(", ")
            )));
        }
        if auto_checkpoint_threshold_files < 1 {
            return Err(EnterpriseError::Validation(format!(
                "auto_checkpoint_threshold_files must be >= 1, got {auto_checkpoint_threshold_files}"
            )));
        }
        if auto_checkpoint_threshold_minutes < 1 {
            return Err(EnterpriseError::Validation(format!(
                "auto_checkpoint_threshold_minutes must be >= 1, got {auto_checkpoint_threshold_minutes}"
            )));
        }
        Ok(Self {
            mode,
            budget_profile,
            pre_commit_verification,
            out_of_scope_warning,
            auto_checkpoint_threshold_files,
            auto_checkpoint_threshold_minutes,
            warn_on_security_summary: true,
        })
    }

    /// `from_config`: typos caen a defaults seguros (nunca falla).
    pub fn from_config_values(
        mode: &str,
        default_budget_profile: &str,
        auto_checkpoint_files: i64,
        auto_checkpoint_minutes: i64,
        clock: Arc<dyn Clock>,
    ) -> Result<Self, EnterpriseError> {
        let parsed_mode = AutopilotMode::parse(mode).unwrap_or(AutopilotMode::Assist);
        let budget = if KNOWN_BUDGET_PROFILES.contains(&default_budget_profile) {
            default_budget_profile.to_string()
        } else {
            DEFAULT_BUDGET_PROFILE.to_string()
        };
        let is_observe = parsed_mode == AutopilotMode::Observe;
        Self::new(
            parsed_mode,
            budget,
            parsed_mode == AutopilotMode::Autopilot,
            !is_observe,
            (auto_checkpoint_files).max(1),
            (auto_checkpoint_minutes).max(1),
            clock,
        )
        .map(|mut p| {
            p.warn_on_security_summary = !is_observe;
            p
        })
    }

    /// `from_config(config)`.
    pub fn from_config(
        config: &AutopilotConfig,
        clock: Arc<dyn Clock>,
    ) -> Result<Self, EnterpriseError> {
        Self::from_config_values(
            &config.mode,
            &config.default_budget_profile,
            config.auto_checkpoint_files,
            config.auto_checkpoint_minutes,
            clock,
        )
    }
}

pub struct PolicyEnforcer {
    policy: AutopilotPolicy,
    clock: Arc<dyn Clock>,
}

impl PolicyEnforcer {
    pub fn new(policy: AutopilotPolicy) -> Self {
        // El enforcer usa el reloj del sistema salvo inyección explícita
        // (with_clock) para tests/gate.
        Self {
            policy,
            clock: Arc::new(cortex_enterprise::clock::SystemClock),
        }
    }

    /// Inyección de reloj para tests/gate.
    pub fn with_clock(mut self, clock: Arc<dyn Clock>) -> Self {
        self.clock = clock;
        self
    }

    pub fn policy(&self) -> &AutopilotPolicy {
        &self.policy
    }

    /// `on_session_open`.
    pub fn on_session_open(
        &self,
        session: &SessionRecord,
        spec_summary: Option<&str>,
    ) -> Vec<EnforcementResult> {
        let mut results = Vec::new();
        if self.policy.mode == AutopilotMode::Observe {
            return results;
        }
        let summary = spec_summary.unwrap_or(&session.spec_summary);
        if self.policy.warn_on_security_summary && looks_security_sensitive(summary) {
            results.push(EnforcementResult::warn(
                "Spec summary mentions security-sensitive terms — review the diff carefully before closing the session.",
            ));
        }
        results
    }

    /// `on_checkpoint`.
    pub fn on_checkpoint(
        &self,
        session: &SessionRecord,
        checkpoint: &Checkpoint,
        files_in_scope: Option<&[String]>,
    ) -> Vec<EnforcementResult> {
        let mut results = Vec::new();
        if self.policy.mode == AutopilotMode::Observe {
            return results;
        }

        if self.policy.out_of_scope_warning {
            if let Some(scope) = files_in_scope {
                let drift: Vec<&String> = checkpoint
                    .artifacts_touched
                    .iter()
                    .filter(|a| !scope.contains(a))
                    .collect();
                if !drift.is_empty() {
                    let mut sorted: Vec<String> = drift.into_iter().cloned().collect();
                    sorted.sort();
                    results.push(EnforcementResult::warn(format!(
                        "Checkpoint touches files outside spec scope: {}",
                        py_list_repr(&sorted)
                    )));
                }
            }
        }

        let since_verified = files_since_last_verified(session);
        if since_verified > self.policy.auto_checkpoint_threshold_files as usize {
            results.push(EnforcementResult::warn(format!(
                "{since_verified} artifact paths touched without a checkpoint that records verified claims"
            )));
        }

        let prior = session.checkpoints.len().saturating_sub(1);
        if prior > 0 && !checkpoint.artifacts_touched.is_empty() {
            let prev_ts: DateTime<Utc> = session.checkpoints[prior - 1].timestamp;
            let elapsed = self.clock.now() - prev_ts;
            let threshold =
                chrono::Duration::minutes(self.policy.auto_checkpoint_threshold_minutes);
            if elapsed > threshold {
                let minutes = elapsed.num_minutes();
                results.push(EnforcementResult::warn(format!(
                    "{minutes} minutes since the previous checkpoint and the new one already touches files — consider checkpointing more often"
                )));
            }
        }
        results
    }

    /// `on_pre_close`.
    pub fn on_pre_close(&self, session: &SessionRecord) -> Vec<EnforcementResult> {
        let mut results = Vec::new();
        let _ = SessionStatus::Open;
        if self.policy.mode == AutopilotMode::Autopilot
            && self.policy.pre_commit_verification
            && !has_verified_checkpoint(session)
        {
            results.push(EnforcementResult::block(
                "Autopilot mode requires at least one checkpoint with verified claims before closing the session (set pre_commit_verification=False to opt out).",
            ));
        }
        results
    }
}

/// `repr(list_de_str_python)`: elementos con comillas simples
/// (`['a', 'b']`), formato canónico de `{sorted(drift)}` en el oráculo.
pub(crate) fn py_list_repr(items: &[String]) -> String {
    let inner: Vec<String> = items.iter().map(|s| format!("'{s}'")).collect();
    format!("[{}]", inner.join(", "))
}
